moviebot/nlu/annotation/joint_bert/joint_bert_train.py

```diff
@@ -0,0 +1,232 @@
+            (intent_logits.argmax(dim=1) == intent_labels).float().mean()
+        )
+        slot_acc = (
+            (
+                (slot_logits.argmax(dim=2) == slot_labels)
+                & (slot_labels != _IGNORE_INDEX)
+            )
+            .float()
+            .mean()
+        )
+
+        # Log the metrics
+        self.log(
+            "val_loss",
+            val_loss,
+            on_step=False,
+            on_epoch=True,
+            prog_bar=True,
+            logger=True,
+        )
+        self.log(
+            "intent_acc",
+            intent_acc,
+            on_step=False,
+            on_epoch=True,
+            prog_bar=True,
+            logger=True,
+        )
+        self.log(
+            "slot_acc",
+            slot_acc,
+            on_step=False,
+            on_epoch=True,
+            prog_bar=True,
+            logger=True,
+        )
+
+        return val_loss
+
+    def configure_optimizers(self):
+        """Configures the optimizer and scheduler for training."""
+        no_decay = ["bias", "LayerNorm.weight"]
+        optimizer_grouped_parameters = [
+            {
+                "params": [
+                    p
+                    for n, p in self.named_parameters()
+                    if not any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": self.hparams.weight_decay,
+            },
+            {
+                "params": [
+                    p
+                    for n, p in self.named_parameters()
+                    if any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": 0.0,
+            },
+        ]
+        optimizer = torch.optim.AdamW(
+            optimizer_grouped_parameters, lr=self.hparams.learning_rate
+        )
+
+        scheduler = get_linear_schedule_with_warmup(
+            optimizer,
+            num_warmup_steps=0,
+            num_training_steps=self.hparams.max_steps,
+        )
+        return [optimizer], [scheduler]
+
+    def save_pretrained(self, path: str) -> None:
+        """Saves the model to the specified directory."""
+
+        if not os.path.exists(path):
+            os.makedirs(path)
+
+        model_path = os.path.join(path, "joint_bert_model.pth")
+        torch.save(self.state_dict(), model_path)
+
+        # Save metadata
+        metadata = {
+            "intents": [intent.name for intent in JointBERTIntent],
+            "slots": [slot.name for slot in JointBERTSlot],
+        }
+        metadata_path = os.path.join(path, "metadata.json")
+        with open(metadata_path, "w") as f:
+            json.dump(metadata, f)
+
+
+def parse_arguments():
+    """Parses the command line arguments."""
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--max_epochs", type=int, default=5)
+    parser.add_argument("--learning_rate", type=float, default=5e-5)
+    parser.add_argument("--weight_decay", type=float, default=0.0)
+    args = parser.parse_args()
```

Comment:
Suggestion to add `--model_output_path` as an argument with `_MODEL_OUTPUT_PATH`.

Reply:
Done